File: QadeerApp.Web/Modules/Administration/TrainingTerm/TrainingTermRow.cs
namespace QadeerApp.Administration;

[ConnectionKey("Default"), Module("Administration"), TableName("TrainingTerms")]
[DisplayName("Training Terms"), InstanceName("Training Term")]
[ReadPermission(PermissionKeys.TrainingTerm.View)]
[ModifyPermission(PermissionKeys.TrainingTerm.Update)]
[InsertPermission(PermissionKeys.TrainingTerm.Insert)]
[UpdatePermission(PermissionKeys.TrainingTerm.Update)]
[DeletePermission(PermissionKeys.TrainingTerm.Delete)]
[LookupScript(Permission = "*")]
public sealed class TrainingTermRow : Serenity.Extensions.Entities.LoggingRow<TrainingTermRow.RowFields>, IIdRow, INameRow, IIsActiveRow, IIsActiveDeletedRow, IDeleteLogRow
{
    [DisplayName("Training Term Id"), Identity, IdProperty]
    public int? TrainingTermId { get => fields.TrainingTermId[this]; set => fields.TrainingTermId[this] = value; }

    [DisplayName("Name"), Size(200), NotNull, QuickSearch, NameProperty]
    public string Name { get => fields.Name[this]; set => fields.Name[this] = value; }

    [DisplayName("Is Active"), NotNull, DefaultValue(1)]
    public short? IsActive { get => fields.IsActive[this]; set => fields.IsActive[this] = value; }

    [Insertable(false), Updatable(false)]
    public int? DeleteUserId { get => fields.DeleteUserId[this]; set => fields.DeleteUserId[this] = value; }

    [Insertable(false), Updatable(false)]
    public DateTime? DeleteDate { get => fields.DeleteDate[this]; set => fields.DeleteDate[this] = value; }

    Int16Field IIsActiveRow.IsActiveField => fields.IsActive;
    Field IDeleteLogRow.DeleteUserIdField => fields.DeleteUserId;
    DateTimeField IDeleteLogRow.DeleteDateField => fields.DeleteDate;

    public class RowFields : Serenity.Extensions.Entities.LoggingRowFields
    {
        public Int32Field TrainingTermId;
        public StringField Name;
        public Int16Field IsActive;
        public Int32Field DeleteUserId;
        public DateTimeField DeleteDate;
    }
}

File: QadeerApp.Web/Modules/ServerTypes/Administration/TrainingTermRow.ts
import { fieldsProxy, getLookup, getLookupAsync } from "@serenity-is/corelib";

export interface TrainingTermRow {
    TrainingTermId?: number;
    Name?: string;
    IsActive?: number;
    DeleteUserId?: number;
    DeleteDate?: string;
    InsertUserId?: number;
    InsertDate?: string;
    UpdateUserId?: number;
    UpdateDate?: string;
}

export abstract class TrainingTermRow {
    static readonly idProperty = 'TrainingTermId';
    static readonly isActiveProperty = 'IsActive';
    static readonly nameProperty = 'Name';
    static readonly localTextPrefix = 'Administration.TrainingTerm';
    static readonly lookupKey = 'Administration.TrainingTerm';

    /** @deprecated use getLookupAsync instead */
    static getLookup() { return getLookup<TrainingTermRow>('Administration.TrainingTerm') }
    static async getLookupAsync() { return getLookupAsync<TrainingTermRow>('Administration.TrainingTerm') }

    static readonly deletePermission = 'Administration:TrainingTerms:Delete';
    static readonly insertPermission = 'Administration:TrainingTerms:Insert';
    static readonly readPermission = 'Administration:TrainingTerms:View';
    static readonly updatePermission = 'Administration:TrainingTerms:Update';

    static readonly Fields = fieldsProxy<TrainingTermRow>();
}
